wmutil.d

```diff
@@ -0,0 +1,162 @@
+/++
+	Cross platform window manager utilities for interacting with other unknown windows on the OS.
+
+	Based on arsd.simpledisplay.
++/
+module arsd.wmutil;
+
+public import arsd.simpledisplay;
+
+static assert(UsingSimpledisplayX11 || UsingSimpledisplayWindows, "wmutil only works on X11 or Windows");
+
+static if (UsingSimpledisplayX11)
+{
+extern(C) nothrow @nogc {
+Atom* XListProperties(Display *display, Window w, int *num_prop_return);
+Status XGetTextProperty(Display *display, Window w, XTextProperty *text_prop_return, Atom property);
+Status XQueryTree(Display *display, Window w, Window *root_return, Window *parent_return, Window **children_return, uint *nchildren_return);
+}
+}
+
+struct WindowChildrenIterator {
+	NativeWindowHandle parent;
+
+	int opApply(int delegate(NativeWindowHandle) dg) const {
+		version (Windows) {
+			struct EnumParams {
+				int result;
+				int delegate(NativeWindowHandle) dg;
+				Exception ex;
+			}
+
+			EnumParams params;
+
+			EnumChildWindows(parent, function (window, lparam) nothrow {
+				EnumParams* args = cast(EnumParams*)lparam;
+				try {
+					args.result = args.dg(window);
+					if (args.result)
+						return 0;
+					else
+						return 1;
+				} catch (Exception e) {
+					args.ex = e;
+					return 0;
+				}
+			}, cast(LPARAM)&params);
+
+			if (params.ex)
+				throw params.ex;
+
+			return params.result;
+		} else static if (UsingSimpledisplayX11) {
+			int result;
+			Window unusedWindow;
+			Window* children;
+			uint numChildren;
+			Status status = XQueryTree(XDisplayConnection.get(), parent, &unusedWindow, &unusedWindow, &children, &numChildren);
+			if (status == 0 || children is null)
+				return 0;
+			scope (exit)
+				XFree(children);
+
+			foreach (window; children[0 .. numChildren]) {
+				result = dg(window);
+				if (result)
+					break;
+			}
+			return result;
+		} else
+			throw new NotYetImplementedException();
+		
+	}
+}
+
+WindowChildrenIterator iterateWindows(NativeWindowHandle parent = NativeWindowHandle.init) {
+	static if (UsingSimpledisplayX11)
+		if (parent == NativeWindowHandle.init)
+			parent = RootWindow(XDisplayConnection.get, DefaultScreen(XDisplayConnection.get));
+
+	return WindowChildrenIterator(parent);
+}
+
+/++
+	Searches for a window with the specified class name and returns the native window handle to it.
+
+	Params:
+		className = the class name to check the window for, case-insensitive.
++/
+version (Windows)
+NativeWindowHandle findWindowByClass(LPCTSTR className) {
```

Comment:
so i do prefer having the signature written once and the static if inside if the interface is cross-platform

Reply:
https://github.com/adamdruppe/arsd/commit/53525181d485fa9a2f2fba4562ad4d86bd545673